:memo: Add a DDEV tutorial (#2325)
platformsh/platformsh-docs: docs/src/development/local/ddev.md

---
title: Use DDEV for local development
sidebarTitle: DDEV
description: Use DDEV to set up local development environments.
weight: 1
---

[DDEV](https://ddev.readthedocs.io/en/stable/) is an open-source tool to get local PHP development environments.
It allows you use Docker in your workflows while maintaining a GitOps workflow.
You get fully containerized environments to run everything locally

> without having to install tools (such as the Platform.sh CLI) on your machine.

Reviewer: ```suggestion
without having to install tools (even the Platform.sh CLI or php or composer) on your machine.
```

This installs DDEV and `ddev`, its command-line interface (CLI).
For more information on `ddev`, run `ddev help`.

## 2. Add DDEV configuration to your project

Get basic configuration set up for your project by running the following command:

```bash
ddev config
```

If you started with a Drupal template, your repository already had DDEV configuration files.
Otherwise, you have to answer a couple of questions about what your repository is
so the correct files are added.

Test the configuration by running the following command:

```bash
ddev start
```

Your project should start running, though without any of the data from your Platform.sh environment.

## 3. Add an API token

To connect DDEV with your Platform.sh account, use a Platform.sh API token.

1. [Create an API token](./../cli/api-tokens.md#get-a-token) in the console.
2. Add the token to your global DDEV configuration:

   ```yaml {location="~/.ddev/global_config.yaml"}
   web_environment:
       - PLATFORMSH_CLI_TOKEN=<API_TOKEN>
   ```

   Note that this isn't the configuration file in your project.
3. Get DDEV to recognize the token by restarting DDEV:

   ```bash
   ddev restart
   ```

> Now you can run `ddev exec platform <command>` from your computer without needing to install the Platform.sh CLI.

Reviewer: A platform DDEV add-on could be a great thing. `ddev get platformsh/platform-setup` could add the .ddev/providers/platform.sh, might even be able to figure out what the env values should be. It could also install a custom command for `ddev platform`.  https://github.com/drud/ddev-addon-template

## 6. (Optional) Install dependencies

If your project has dependencies, you can install them with Composer.
Using the DDEV version of Composer means you don't need to install Composer on your machine.
You can also use different Composer versions in different projects based on their DDEV configurations.

Run the following command:

```bash
ddev composer install
```

If you have any other commands in your [build hook](../../create-apps/hooks/_index.md),
you need to run those explicitly.
You can also add them as [DDEV hooks in your configuration](https://ddev.readthedocs.io/en/stable/users/extending-commands/).

## 7. Run your project

Now your project is ready to run:

```bash
ddev start
```

The command returns the project URL `https://<PROJECT_NAME>.ddev.site/`
as well as a specific port on `https://127.0.0.1`.
Open one of the URLs to see your project running.

## What's next

You've got a project running on a local web server.
Now you can add customizations.
For more ideas and options, see the [DDEV documentation](https://ddev.readthedocs.io/en/stable/).

### Add environment variables

If your project requires environment variables for its hooks or runtime,

> add them to the DDEV environment:

Reviewer 2: :nail_care: ▶️
As opposed to the global DDEV environment setting in [3. Add an API token](https://pr-2325-r3lvrhi-652soceglkw4u.eu-3.platformsh.site/development/local/ddev.html#3-add-an-api-token).
```suggestion
add them to the local DDEV environment:
```

Author: I can see how it would help to clarify, but given that "local" is true for both (for everything DDEV), I'll try to find a different qualifier. :thinking:
